### PR TITLE
Add MinGW (GCC 8)

### PR DESCRIPTION
**Pull request by uilianries**
MinGW is not deprecated in Conan Community it's still well documented in Conan docs and well used on Conan blog.

This PR brings only GCC8, if someone requires older versions, we can add in the future.

Here is a challenge, as CCI doesn't run GCC as compiler, we won't generate packages for this projects.
Or we will need to change it as options,
Or CCI will need to generate it as an internal exception,
Or we just post the code but don't provide any package

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/mingw/all/conandata.yml [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+            sjlj:
+              url: "http://sourceforge.net/projects/mingw-w64/files/Toolchains%20targetting%20Win64/Personal%20Builds/mingw-builds/8.1.0/threads-posix/sjlj/x86_64-8.1.0-release-posix-sjlj-rt_v6-rev0.7z"
+              sha256: "0e994b33dc92576ba92377e2c18c0f96a04c684fa2d7a3546890819964275a1f"
+          win32:
+            seh:
+              url: "http://sourceforge.net/projects/mingw-w64/files/Toolchains%20targetting%20Win64/Personal%20Builds/mingw-builds/8.1.0/threads-win32/seh/x86_64-8.1.0-release-win32-seh-rt_v6-rev0.7z"
+              sha256: "73797b9a1d0b007dbdd6f010e6064dc0aa4df835b042f6947f73a3bf9cc348c5"
+            sjlj:
+              url: "https://sourceforge.net/projects/mingw-w64/files/Toolchains%20targetting%20Win64/Personal%20Builds/mingw-builds/7.3.0/threads-win32/sjlj/x86_64-7.3.0-release-win32-sjlj-rt_v5-rev0.7z"
+              sha256: "f5d3add9afe6ef0b8ff07250076aa1410f238bcbf15b45c245a06cbf7111e645"
```
Comment (ericLemanissier) [inline recipes/mingw/all/conandata.yml:19]: this link points to 7.3
```suggestion
              url: "https://sourceforge.net/projects/mingw-w64/files/Toolchains%20targetting%20Win64/Personal%20Builds/mingw-builds/8.1.0/threads-win32/sjlj/x86_64-8.1.0-release-win32-sjlj-rt_v6-rev0.7z"
              sha256: "?"
```

**Comment by ericLemanissier** [comment]
Will the bot upload the recipe to bintray even if all configurations are invalid ?

File: recipes/mingw/all/conandata.yml (hunk 1 above)
Comment (uilianries) [reply recipes/mingw/all/conandata.yml:19]: CtrlC+CtrlV 🤦

**Comment by uilianries** [comment]
> Will the bot upload the recipe to bintray even if all configurations are invalid ?

No, it would require an internal feature.

File: recipes/mingw/all/conandata.yml (hunk 1 above)
Comment (uilianries) [reply recipes/mingw/all/conandata.yml]: Done!

**Comment by dmn-star** [comment]
@uilianries Are there any particular reasons why you didn't use tools.unzip?

**Comment by uilianries** [comment]
@dmn-star [tools.unzip](https://docs.conan.io/en/latest/reference/tools.html#tools-unzip) does not support .7z

7z requires a specific package (7z installer), but we don't export any helper, like python requires.

**Comment by dmn-star** [comment]
thx

File: recipes/mingw/all/conandata.yml [hunk 2]
```diff
@@ -0,0 +1,18 @@
+sources:
+  "8.1":
```
Comment (SpaceIm) [inline recipes/mingw/all/conandata.yml:2]: MinGW version and gcc version are two different things. What's the version of MinGW-w64 here?

Comment (uilianries) [reply recipes/mingw/all/conandata.yml:2]: version : MinGW-W64-builds-4.3.5
user    : nixman
date    : 05.12.2018-10:29:36 AM

File: recipes/mingw/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+class MingwConan(ConanFile):
+    name = "mingw"
+    description = "MinGW is a contraction of Minimalist GNU for Windows"
+    url = "https://github.com/conan-io/conan-center-index"
+    homepage = "https://www.mingw.org"
+    license = "ZPL-2.1", "MIT", "GPL-2.0-or-later"
+    topics = ("gcc", "gnu", "unix", "mingw32", "binutils")
+    settings = "os", "arch"
+    options = {"threads": ["posix", "win32"], "exception": ["seh", "sjlj"]}
```
Comment (ericLemanissier) [inline recipes/mingw/all/conanfile.py:14]: This means we allow option mingw:threads and mingw:exception to be different from settings compiler.threads and compiler.exception. This sounds very dangerous for the user

Comment (uilianries) [reply recipes/mingw/all/conanfile.py:14]: Yes, but unfortunately this is the only way right now. We are using the same approach for Android NDK. With new profiles approach (build x target) we still can't solve it, but it's listed to be solve on Conan 2.0.

The recommendation will be generate a profile which combines both settings and options.

Comment (ericLemanissier) [reply recipes/mingw/all/conanfile.py:14]: Which are NDK are you referring to ? https://github.com/conan-io/conan-center-index/blob/master/recipes/android-ndk/all/conanfile.py does not seem to use options.
At the very least we should print a warning if the self.settings.compiler is not "gcc", and if one of the options does not match the settings

File: recipes/mingw/all/conandata.yml (hunk 2 above)
Comment (ericLemanissier) [reply recipes/mingw/all/conandata.yml:2]: nixman's personal builds are [not maintained any more by the way](https://sourceforge.net/p/mingw-w64/discussion/723797/thread/ea9a5b00fb/). We probably want to use something more future-proof, for example http://winlibs.com/

Comment (SpaceIm) [reply recipes/mingw/all/conandata.yml:2]: Agree, but it provides only posix/seh for 64bits, and posix/dwark for 32bits

Comment (uilianries) [reply recipes/mingw/all/conandata.yml:2]: Yes, only few configurations available.

File: recipes/mingw/all/conanfile.py (hunk 3 above)
Comment (uilianries) [reply recipes/mingw/all/conanfile.py:14]: Yes, a warning will be made

File: recipes/mingw/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    def validate(self):
+        if self.settings.os != "Windows":
+            raise ConanInvalidConfiguration("MinGW is only supported by Windows.")
+        if self.settings.compiler == "gcc":
+            if str(self.settings.compiler.threads) != str(self.options.threads):
+                self.output.warn("MinGW threads may not differ from 'settings.compiler.threads'.")
+            if str(self.settings.compiler.exception) != str(self.options.exception):
+                self.output.warn("MinGW exception may not differ from 'settings.compiler.exception'.")
+        else:
+            self.output.warn("Only 'gcc' should be used as compiler.")
```
Comment (madebr) [inline recipes/mingw/all/conanfile.py:28]: Aren't we mixing build profile and host profile here?

This recipe will most often be used as a build requirement ==> used in the build context.
So the conan host profile should be checked for correct settings.
Looking at the [cross building documentatino](https://docs.conan.io/en/latest/systems_cross_building/cross_building.html), shouldn't `self.settings_target` be used?

File: recipes/mingw/all/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        del self.info.settings.compiler
+
+    def package_info(self):
+        bin_path = os.path.join(self.package_folder, "bin")
+        self.output.info("Appending PATH env var with : {}".format(bin_path))
+        self.env_info.PATH.append(bin_path)
+        self.env_info.MINGW_HOME = str(self.package_folder)
+        self.env_info.CONAN_CMAKE_GENERATOR = "MinGW Makefiles"
+        self.env_info.CXX = os.path.join(self.package_folder, "bin", "g++.exe").replace("\\", "/")
+        self.env_info.CC = os.path.join(self.package_folder, "bin", "gcc.exe").replace("\\", "/")
```
Comment (madebr) [inline recipes/mingw/all/conanfile.py:54]: Should it set additional variables?
e.g. `NM`, `RANIB`, `STRINGS`, `OBJDUMP`, ...

File: recipes/mingw/all/test_package/conanfile.py [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+from conans import ConanFile, tools
+
+
+class MinGWTestConan(ConanFile):
+    generators = "gcc"
+    settings = "os", "arch", "compiler", "build_type"
+
+    def build(self):
+        source_file = os.path.join(self.source_folder, "main.cpp")
+        self.run("gcc {} @conanbuildinfo.gcc -lstdc++ -o main".format(source_file), run_environment=True)
```
Comment (madebr) [inline recipes/mingw/all/test_package/conanfile.py:11]: I haven't checked this recipe locally, but does this package contain `gcc`?
Most often, mingw compilers have a prefix: e.g. `x86_64-w64-mingw32-gcc` or something similar.

File: recipes/mingw/all/conanfile.py (hunk 3 above)
Comment (jgsogo) [reply recipes/mingw/all/conanfile.py:14]: We can add a check in the package_info when using two-profiles approach:

```python

def package_info(self):
    if hasattr(self, 'settings_target'):
        assert self.settings_target.compiler == 'gcc'
        if self.settings_target.compiler.threads != self.options.threads:
            raise ConanInvalidConfiguration('Build requires 'mingw' provides binaries for gcc with threads={}, your profile:host declares threads={}, please....')
        ....

```

I don't know how friendly the exception will be, but at least the message should tell the user information about the mismatch in the settings and options.

Comment (uilianries) [reply recipes/mingw/all/conanfile.py:14]: That's what I was looking for!

File: recipes/mingw/all/conanfile.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
+import os
+from conans import ConanFile, tools
+from conans.errors import ConanInvalidConfiguration
+
+
+class MingwConan(ConanFile):
+    name = "mingw"
+    description = "MinGW is a contraction of Minimalist GNU for Windows"
+    url = "https://github.com/conan-io/conan-center-index"
+    homepage = "https://www.mingw.org"
```
Comment (Croydon) [inline recipes/mingw/all/conanfile.py:10]: This site does not load for me

File: recipes/mingw/all/conandata.yml (hunk 2 above)
Comment (Croydon) [reply recipes/mingw/all/conandata.yml:2]: What about http://mingw-w64.org?

File: recipes/mingw/all/conanfile.py (hunk 7 above)
Comment (uilianries) [reply recipes/mingw/all/conanfile.py:10]: I'll move to http://mingw-w64.org/doku.php

Comment (uilianries) [inline recipes/mingw/all/conanfile.py:10]:
```suggestion
    homepage = "http://mingw-w64.org/doku.php"
```

**Comment by uilianries** [comment]
It should work now https://github.com/conan-io/hooks/pull/267

File: recipes/mingw/all/conanfile.py [hunk 8]
```diff
@@ -0,0 +1,73 @@
+import os
+from conans import ConanFile, tools
+from conans.errors import ConanInvalidConfiguration
+
+
+class MingwConan(ConanFile):
+    name = "mingw"
```
Comment (madebr) [inline recipes/mingw/all/conanfile.py:7]: This is a recipe for a fork of mingw, mingw-w64 (see wikipedia: https://en.wikipedia.org/wiki/MinGW)
So I propose to rename it to `mingw-w64`.
```suggestion
    name = "mingw-w64"
```

Comment (ericLemanissier) [reply recipes/mingw/all/conanfile.py:7]: especially with MinGW coming back to life on https://osdn.net/projects/mingw/

Comment (uilianries) [reply recipes/mingw/all/conanfile.py:7]: Great find! I'll update it, thanks

Comment (uilianries) [reply recipes/mingw/all/conanfile.py]: Done

File: recipes/mingw/all/conandata.yml (hunk 2 above)
Comment (SpaceIm) [reply recipes/mingw/all/conandata.yml]: @uilianries For me, it's not resolved, the version is misleading. When I see mingw/8 recipe, I don't expect to have MinGW 4.3.5 ;)

**Comment by uilianries** [comment]
Requires https://github.com/conan-io/hooks/pull/288

**Review by danimtb — APPROVED** [review]

**Review by prince-chrismc — APPROVED** [review]

**Comment by SpaceIm** [comment]
@uilianries I've left a comment https://github.com/conan-io/conan-center-index/pull/4327#discussion_r609608837.

**Review by SSE4 — APPROVED** [review]

**Pull request merged**